// @medusajs/utils/dist/dal/mikro-orm/mikro-orm-repository.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MikroOrmBaseTreeRepository = exports.MikroOrmBaseRepository = exports.MikroOrmAbstractTreeRepositoryBase = exports.MikroOrmAbstractBaseRepository = exports.MikroOrmBase = void 0;
var common_1 = require("../../common");
var decorators_1 = require("../../decorators");
var modules_sdk_1 = require("../../modules-sdk");
var utils_1 = require("../utils");
var utils_2 = require("./utils");
var MikroOrmBase = /** @class */ (function () {
    function MikroOrmBase(_a) {
        var manager = _a.manager;
        this.manager_ = manager;
    }
    MikroOrmBase.prototype.getFreshManager = function () {
        return (this.manager_.fork
            ? this.manager_.fork()
            : this.manager_);
    };
    MikroOrmBase.prototype.getActiveManager = function (_a) {
        var _b;
        var _c = _a === void 0 ? {} : _a, transactionManager = _c.transactionManager, manager = _c.manager;
        return ((_b = transactionManager !== null && transactionManager !== void 0 ? transactionManager : manager) !== null && _b !== void 0 ? _b : this.manager_);
    };
    MikroOrmBase.prototype.transaction = function (task, options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, utils_1.transactionWrapper.bind(this)(task, options)];
                    case 1: 
                    // @ts-ignore
                    return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    MikroOrmBase.prototype.serialize = function (data, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, utils_2.mikroOrmSerializer)(data, options)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    var _a;
    __decorate([
        __param(0, (0, decorators_1.MedusaContext)()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", typeof (_a = typeof TManager !== "undefined" && TManager) === "function" ? _a : Object)
    ], MikroOrmBase.prototype, "getActiveManager", null);
    return MikroOrmBase;
}());
exports.MikroOrmBase = MikroOrmBase;
var MikroOrmAbstractBaseRepository = /** @class */ (function (_super) {
    __extends(MikroOrmAbstractBaseRepository, _super);
    function MikroOrmAbstractBaseRepository() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MikroOrmAbstractBaseRepository.prototype.update = function (data, context) {
        throw new Error("Method not implemented.");
    };
    MikroOrmAbstractBaseRepository.prototype.softDelete = function (idsOrFilter, _a) {
        var _b = _a === void 0 ? {} : _a, manager = _b.transactionManager;
        return __awaiter(this, void 0, void 0, function () {
            var isArray, filter, entities, date, softDeletedEntitiesMap;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        isArray = Array.isArray(idsOrFilter);
                        filter = isArray || (0, common_1.isString)(idsOrFilter)
                            ? {
                                id: {
                                    $in: isArray ? idsOrFilter : [idsOrFilter],
                                },
                            }
                            : idsOrFilter;
                        return [4 /*yield*/, this.find({ where: filter })];
                    case 1:
                        entities = _c.sent();
                        date = new Date();
                        return [4 /*yield*/, (0, utils_2.mikroOrmUpdateDeletedAtRecursively)(manager, entities, date)];
                    case 2:
                        _c.sent();
                        softDeletedEntitiesMap = (0, utils_1.getSoftDeletedCascadedEntitiesIdsMappedBy)({
                            entities: entities,
                        });
                        return [2 /*return*/, [entities, softDeletedEntitiesMap]];
                }
            });
        });
    };
    MikroOrmAbstractBaseRepository.prototype.restore = function (idsOrFilter, _a) {
        var _b = _a === void 0 ? {} : _a, manager = _b.transactionManager;
        return __awaiter(this, void 0, void 0, function () {
            var isArray, filter, query, entities, softDeletedEntitiesMap;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        isArray = Array.isArray(idsOrFilter);
                        filter = isArray || (0, common_1.isString)(idsOrFilter)
                            ? {
                                id: {
                                    $in: isArray ? idsOrFilter : [idsOrFilter],
                                },
                            }
                            : idsOrFilter;
                        query = (0, modules_sdk_1.buildQuery)(filter, {
                            withDeleted: true,
                        });
                        return [4 /*yield*/, this.find(query)];
                    case 1:
                        entities = _c.sent();
                        return [4 /*yield*/, (0, utils_2.mikroOrmUpdateDeletedAtRecursively)(manager, entities, null)];
                    case 2:
                        _c.sent();
                        softDeletedEntitiesMap = (0, utils_1.getSoftDeletedCascadedEntitiesIdsMappedBy)({
                            entities: entities,
                            restored: true,
                        });
                        return [2 /*return*/, [entities, softDeletedEntitiesMap]];
                }
            });
        });
    };
    MikroOrmAbstractBaseRepository.prototype.applyFreeTextSearchFilters = function (findOptions, retrieveConstraintsToApply) {
        if (!("q" in findOptions.where) || !findOptions.where.q) {
            return;
        }
        var q = findOptions.where.q;
        delete findOptions.where.q;
        findOptions.where = {
            $and: [findOptions.where, { $or: retrieveConstraintsToApply(q) }],
        };
    };
    __decorate([
        (0, modules_sdk_1.InjectTransactionManager)(),
        __param(1, (0, decorators_1.MedusaContext)()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Object]),
        __metadata("design:returntype", Promise)
    ], MikroOrmAbstractBaseRepository.prototype, "softDelete", null);
    __decorate([
        (0, modules_sdk_1.InjectTransactionManager)(),
        __param(1, (0, decorators_1.MedusaContext)()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Object]),
        __metadata("design:returntype", Promise)
    ], MikroOrmAbstractBaseRepository.prototype, "restore", null);
    return MikroOrmAbstractBaseRepository;
}(MikroOrmBase));
exports.MikroOrmAbstractBaseRepository = MikroOrmAbstractBaseRepository;
var MikroOrmAbstractTreeRepositoryBase = /** @class */ (function (_super) {
    __extends(MikroOrmAbstractTreeRepositoryBase, _super);
    function MikroOrmAbstractTreeRepositoryBase(_a) {
        var manager = _a.manager;
        return _super.apply(this, __spreadArray([], __read(arguments), false)) || this;
    }
    return MikroOrmAbstractTreeRepositoryBase;
}(MikroOrmBase));
exports.MikroOrmAbstractTreeRepositoryBase = MikroOrmAbstractTreeRepositoryBase;
/**
 * Privileged extends of the abstract classes unless most of the methods can't be implemented
 * in your repository. This base repository is also used to provide a base repository
 * injection if needed to be able to use the common methods without being related to an entity.
 * In this case, none of the method will be implemented except the manager and transaction
 * related ones.
 */
var MikroOrmBaseRepository = /** @class */ (function (_super) {
    __extends(MikroOrmBaseRepository, _super);
    function MikroOrmBaseRepository(_a) {
        var manager = _a.manager;
        return _super.apply(this, __spreadArray([], __read(arguments), false)) || this;
    }
    MikroOrmBaseRepository.prototype.create = function (data, context) {
        throw new Error("Method not implemented.");
    };
    MikroOrmBaseRepository.prototype.update = function (data, context) {
        throw new Error("Method not implemented.");
    };
    MikroOrmBaseRepository.prototype.delete = function (ids, context) {
        throw new Error("Method not implemented.");
    };
    MikroOrmBaseRepository.prototype.find = function (options, context) {
        throw new Error("Method not implemented.");
    };
    MikroOrmBaseRepository.prototype.findAndCount = function (options, context) {
        throw new Error("Method not implemented.");
    };
    return MikroOrmBaseRepository;
}(MikroOrmAbstractBaseRepository));
exports.MikroOrmBaseRepository = MikroOrmBaseRepository;
var MikroOrmBaseTreeRepository = /** @class */ (function (_super) {
    __extends(MikroOrmBaseTreeRepository, _super);
    function MikroOrmBaseTreeRepository(_a) {
        var manager = _a.manager;
        return _super.apply(this, __spreadArray([], __read(arguments), false)) || this;
    }
    MikroOrmBaseTreeRepository.prototype.find = function (options, transformOptions, context) {
        throw new Error("Method not implemented.");
    };
    MikroOrmBaseTreeRepository.prototype.findAndCount = function (options, transformOptions, context) {
        throw new Error("Method not implemented.");
    };
    MikroOrmBaseTreeRepository.prototype.create = function (data, context) {
        throw new Error("Method not implemented.");
    };
    MikroOrmBaseTreeRepository.prototype.delete = function (id, context) {
        throw new Error("Method not implemented.");
    };
    return MikroOrmBaseTreeRepository;
}(MikroOrmAbstractTreeRepositoryBase));
exports.MikroOrmBaseTreeRepository = MikroOrmBaseTreeRepository;
