// @medusajs/medusa/dist/commands/migrate.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var awilix_1 = require("awilix");
var get_migrations_1 = __importStar(require("./utils/get-migrations"));
var utils_1 = require("@medusajs/utils");
var config_1 = __importDefault(require("../loaders/config"));
var database_1 = __importDefault(require("../loaders/database"));
var feature_flags_1 = __importDefault(require("../loaders/feature-flags"));
var logger_1 = __importDefault(require("../loaders/logger"));
var medusa_app_1 = require("../loaders/medusa-app");
var pg_connection_1 = __importDefault(require("../loaders/pg-connection"));
var getDataSource = function (directory) { return __awaiter(void 0, void 0, void 0, function () {
    var configModule, featureFlagRouter, coreMigrations, moduleMigrations, container;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                configModule = (0, config_1.default)(directory);
                featureFlagRouter = (0, feature_flags_1.default)(configModule);
                coreMigrations = (0, get_migrations_1.default)(directory, featureFlagRouter).coreMigrations;
                moduleMigrations = (0, get_migrations_1.getModuleSharedResources)(configModule, featureFlagRouter).migrations;
                container = (0, awilix_1.createContainer)();
                container.register("db_entities", (0, awilix_1.asValue)([]));
                return [4 /*yield*/, (0, database_1.default)({
                        container: container,
                        configModule: configModule,
                        customOptions: {
                            migrations: coreMigrations.concat(moduleMigrations),
                            logging: "all",
                        },
                    })];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
var runLinkMigrations = function (directory) { return __awaiter(void 0, void 0, void 0, function () {
    var configModule, container, runMigrations, options;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                configModule = (0, config_1.default)(directory);
                container = (0, utils_1.createMedusaContainer)();
                return [4 /*yield*/, (0, pg_connection_1.default)({ configModule: configModule, container: container })];
            case 1:
                _a.sent();
                return [4 /*yield*/, (0, medusa_app_1.loadMedusaApp)({ configModule: configModule, container: container }, { registerInContainer: false })];
            case 2:
                runMigrations = (_a.sent()).runMigrations;
                options = {
                    database: {
                        clientUrl: configModule.projectConfig.database_url,
                    },
                };
                return [4 /*yield*/, runMigrations(options)];
            case 3:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
var main = function (_a) {
    var directory = _a.directory;
    return __awaiter(this, void 0, void 0, function () {
        var args, configModule, dataSource, featureFlagRouter, unapplied;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    args = process.argv;
                    args.shift();
                    args.shift();
                    args.shift();
                    configModule = (0, config_1.default)(directory);
                    return [4 /*yield*/, getDataSource(directory)];
                case 1:
                    dataSource = _b.sent();
                    featureFlagRouter = (0, feature_flags_1.default)(configModule);
                    if (!(args[0] === "run")) return [3 /*break*/, 7];
                    return [4 /*yield*/, dataSource.runMigrations()];
                case 2:
                    _b.sent();
                    return [4 /*yield*/, dataSource.destroy()];
                case 3:
                    _b.sent();
                    return [4 /*yield*/, (0, get_migrations_1.runIsolatedModulesMigration)(configModule)];
                case 4:
                    _b.sent();
                    if (!featureFlagRouter.isFeatureEnabled(utils_1.MedusaV2Flag.key)) return [3 /*break*/, 6];
                    return [4 /*yield*/, runLinkMigrations(directory)];
                case 5:
                    _b.sent();
                    _b.label = 6;
                case 6:
                    process.exit();
                    logger_1.default.info("Migrations completed.");
                    return [3 /*break*/, 14];
                case 7:
                    if (!(args[0] === "revert")) return [3 /*break*/, 11];
                    return [4 /*yield*/, dataSource.undoLastMigration({ transaction: "all" })];
                case 8:
                    _b.sent();
                    return [4 /*yield*/, dataSource.destroy()];
                case 9:
                    _b.sent();
                    return [4 /*yield*/, (0, get_migrations_1.revertIsolatedModulesMigration)(configModule)];
                case 10:
                    _b.sent();
                    logger_1.default.info("Migrations reverted.");
                    return [3 /*break*/, 14];
                case 11:
                    if (!(args[0] === "show")) return [3 /*break*/, 14];
                    return [4 /*yield*/, dataSource.showMigrations()];
                case 12:
                    unapplied = _b.sent();
                    logger_1.default.info(unapplied);
                    return [4 /*yield*/, dataSource.destroy()];
                case 13:
                    _b.sent();
                    process.exit(unapplied ? 1 : 0);
                    _b.label = 14;
                case 14: return [2 /*return*/];
            }
        });
    });
};
exports.default = main;
