// medusa-payment-stripe/dist/core/__fixtures__/data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updatePaymentDataWithoutAmountData = exports.updatePaymentDataWithAmountData = exports.updatePaymentContextFailWithDifferentAmount = exports.updatePaymentContextWithDifferentAmount = exports.updatePaymentContextWithWrongEmail = exports.updatePaymentContextWithExistingCustomerStripeId = exports.updatePaymentContextWithExistingCustomer = exports.retrievePaymentFailData = exports.retrievePaymentSuccessData = exports.refundPaymentFailData = exports.refundPaymentSuccessData = exports.deletePaymentPartiallyFailData = exports.deletePaymentFailData = exports.deletePaymentSuccessData = exports.capturePaymentContextPartiallyFailData = exports.capturePaymentContextFailData = exports.capturePaymentContextSuccessData = exports.cancelPaymentPartiallyFailData = exports.cancelPaymentFailData = exports.cancelPaymentSuccessData = exports.authorizePaymentSuccessData = exports.initiatePaymentContextWithFailIntentCreation = exports.initiatePaymentContextWithWrongEmail = exports.initiatePaymentContextWithExistingCustomerStripeId = exports.initiatePaymentContextWithExistingCustomer = void 0;
var stripe_1 = require("../../__mocks__/stripe");
var data_1 = require("../../__fixtures__/data");
// INITIATE PAYMENT DATA
exports.initiatePaymentContextWithExistingCustomer = {
    email: stripe_1.EXISTING_CUSTOMER_EMAIL,
    currency_code: "usd",
    amount: 1000,
    resource_id: "test",
    customer: {},
    context: {},
    paymentSessionData: {},
};
exports.initiatePaymentContextWithExistingCustomerStripeId = {
    email: stripe_1.EXISTING_CUSTOMER_EMAIL,
    currency_code: "usd",
    amount: 1000,
    resource_id: "test",
    customer: {
        metadata: {
            stripe_id: "test",
        },
    },
    context: {},
    paymentSessionData: {},
};
exports.initiatePaymentContextWithWrongEmail = {
    email: stripe_1.WRONG_CUSTOMER_EMAIL,
    currency_code: "usd",
    amount: 1000,
    resource_id: "test",
    customer: {},
    context: {},
    paymentSessionData: {},
};
exports.initiatePaymentContextWithFailIntentCreation = {
    email: stripe_1.EXISTING_CUSTOMER_EMAIL,
    currency_code: "usd",
    amount: 1000,
    resource_id: "test",
    customer: {},
    context: {
        payment_description: "fail",
    },
    paymentSessionData: {},
};
// AUTHORIZE PAYMENT DATA
exports.authorizePaymentSuccessData = {
    id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
};
// CANCEL PAYMENT DATA
exports.cancelPaymentSuccessData = {
    id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
};
exports.cancelPaymentFailData = {
    id: stripe_1.FAIL_INTENT_ID,
};
exports.cancelPaymentPartiallyFailData = {
    id: stripe_1.PARTIALLY_FAIL_INTENT_ID,
};
// CAPTURE PAYMENT DATA
exports.capturePaymentContextSuccessData = {
    paymentSessionData: {
        id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
    },
};
exports.capturePaymentContextFailData = {
    paymentSessionData: {
        id: stripe_1.FAIL_INTENT_ID,
    },
};
exports.capturePaymentContextPartiallyFailData = {
    paymentSessionData: {
        id: stripe_1.PARTIALLY_FAIL_INTENT_ID,
    },
};
// DELETE PAYMENT DATA
exports.deletePaymentSuccessData = {
    id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
};
exports.deletePaymentFailData = {
    id: stripe_1.FAIL_INTENT_ID,
};
exports.deletePaymentPartiallyFailData = {
    id: stripe_1.PARTIALLY_FAIL_INTENT_ID,
};
// REFUND PAYMENT DATA
exports.refundPaymentSuccessData = {
    id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
};
exports.refundPaymentFailData = {
    id: stripe_1.FAIL_INTENT_ID,
};
// RETRIEVE PAYMENT DATA
exports.retrievePaymentSuccessData = {
    id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
};
exports.retrievePaymentFailData = {
    id: stripe_1.FAIL_INTENT_ID,
};
// UPDATE PAYMENT DATA
exports.updatePaymentContextWithExistingCustomer = {
    email: stripe_1.EXISTING_CUSTOMER_EMAIL,
    currency_code: "usd",
    amount: 1000,
    resource_id: "test",
    customer: {},
    context: {},
    paymentSessionData: {
        customer: "test",
        amount: 1000,
    },
};
exports.updatePaymentContextWithExistingCustomerStripeId = {
    email: stripe_1.EXISTING_CUSTOMER_EMAIL,
    currency_code: "usd",
    amount: 1000,
    resource_id: "test",
    customer: {
        metadata: {
            stripe_id: "test",
        },
    },
    context: {},
    paymentSessionData: {
        customer: "test",
        amount: 1000,
    },
};
exports.updatePaymentContextWithWrongEmail = {
    email: stripe_1.WRONG_CUSTOMER_EMAIL,
    currency_code: "usd",
    amount: 1000,
    resource_id: "test",
    customer: {},
    context: {},
    paymentSessionData: {
        customer: "test",
        amount: 1000,
    },
};
exports.updatePaymentContextWithDifferentAmount = {
    email: stripe_1.WRONG_CUSTOMER_EMAIL,
    currency_code: "usd",
    amount: 2000,
    resource_id: "test",
    customer: {
        metadata: {
            stripe_id: "test",
        },
    },
    context: {},
    paymentSessionData: {
        id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
        customer: "test",
        amount: 1000,
    },
};
exports.updatePaymentContextFailWithDifferentAmount = {
    email: stripe_1.WRONG_CUSTOMER_EMAIL,
    currency_code: "usd",
    amount: 2000,
    resource_id: "test",
    customer: {
        metadata: {
            stripe_id: "test",
        },
    },
    context: {
        metadata: {
            stripe_id: "test",
        },
    },
    paymentSessionData: {
        id: stripe_1.FAIL_INTENT_ID,
        customer: "test",
        amount: 1000,
    },
};
exports.updatePaymentDataWithAmountData = {
    sessionId: stripe_1.STRIPE_ID,
    amount: 2000,
};
exports.updatePaymentDataWithoutAmountData = {
    sessionId: stripe_1.STRIPE_ID,
    customProp: "test",
};
