// medusa-payment-stripe/dist/core/__tests__/stripe-base.spec.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var os_1 = require("os");
var stripe_test_1 = require("../__fixtures__/stripe-test");
var data_1 = require("../../__fixtures__/data");
var medusa_1 = require("@medusajs/medusa");
var data_2 = require("../__fixtures__/data");
var stripe_1 = require("../../__mocks__/stripe");
var types_1 = require("../../types");
var container = {};
describe.skip("StripeTest", function () {
    describe("getPaymentStatus", function () {
        var _this = this;
        var stripeTest;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        scopedContainer = __assign({}, container);
                        stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                        return [4 /*yield*/, stripeTest.init()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should return the correct status", function () { return __awaiter(_this, void 0, void 0, function () {
            var status;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.getPaymentStatus({
                            id: data_1.PaymentIntentDataByStatus.REQUIRES_PAYMENT_METHOD.id,
                        })];
                    case 1:
                        status = _a.sent();
                        expect(status).toBe(medusa_1.PaymentSessionStatus.PENDING);
                        return [4 /*yield*/, stripeTest.getPaymentStatus({
                                id: data_1.PaymentIntentDataByStatus.REQUIRES_CONFIRMATION.id,
                            })];
                    case 2:
                        status = _a.sent();
                        expect(status).toBe(medusa_1.PaymentSessionStatus.PENDING);
                        return [4 /*yield*/, stripeTest.getPaymentStatus({
                                id: data_1.PaymentIntentDataByStatus.PROCESSING.id,
                            })];
                    case 3:
                        status = _a.sent();
                        expect(status).toBe(medusa_1.PaymentSessionStatus.PENDING);
                        return [4 /*yield*/, stripeTest.getPaymentStatus({
                                id: data_1.PaymentIntentDataByStatus.REQUIRES_ACTION.id,
                            })];
                    case 4:
                        status = _a.sent();
                        expect(status).toBe(medusa_1.PaymentSessionStatus.REQUIRES_MORE);
                        return [4 /*yield*/, stripeTest.getPaymentStatus({
                                id: data_1.PaymentIntentDataByStatus.CANCELED.id,
                            })];
                    case 5:
                        status = _a.sent();
                        expect(status).toBe(medusa_1.PaymentSessionStatus.CANCELED);
                        return [4 /*yield*/, stripeTest.getPaymentStatus({
                                id: data_1.PaymentIntentDataByStatus.REQUIRES_CAPTURE.id,
                            })];
                    case 6:
                        status = _a.sent();
                        expect(status).toBe(medusa_1.PaymentSessionStatus.AUTHORIZED);
                        return [4 /*yield*/, stripeTest.getPaymentStatus({
                                id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
                            })];
                    case 7:
                        status = _a.sent();
                        expect(status).toBe(medusa_1.PaymentSessionStatus.AUTHORIZED);
                        return [4 /*yield*/, stripeTest.getPaymentStatus({
                                id: "unknown-id",
                            })];
                    case 8:
                        status = _a.sent();
                        expect(status).toBe(medusa_1.PaymentSessionStatus.PENDING);
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe("initiatePayment", function () {
        var _this = this;
        var stripeTest;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                scopedContainer = __assign({}, container);
                stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                return [2 /*return*/];
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should succeed with an existing customer but no stripe id", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.initiatePayment(data_2.initiatePaymentContextWithExistingCustomer)];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalledWith({
                            email: data_2.initiatePaymentContextWithExistingCustomer.email,
                        });
                        expect(stripe_1.StripeMock.paymentIntents.create).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.paymentIntents.create).toHaveBeenCalledWith(expect.objectContaining({
                            description: undefined,
                            amount: data_2.initiatePaymentContextWithExistingCustomer.amount,
                            currency: data_2.initiatePaymentContextWithExistingCustomer.currency_code,
                            metadata: {
                                resource_id: data_2.initiatePaymentContextWithExistingCustomer.resource_id,
                            },
                            capture_method: "manual",
                        }));
                        expect(result).toEqual(expect.objectContaining({
                            session_data: expect.any(Object),
                            update_requests: {
                                customer_metadata: {
                                    stripe_id: stripe_1.STRIPE_ID,
                                },
                            },
                        }));
                        return [2 /*return*/];
                }
            });
        }); });
        it("should succeed with an existing customer with an existing stripe id", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.initiatePayment(data_2.initiatePaymentContextWithExistingCustomerStripeId)];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.customers.create).not.toHaveBeenCalled();
                        expect(stripe_1.StripeMock.paymentIntents.create).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.paymentIntents.create).toHaveBeenCalledWith(expect.objectContaining({
                            description: undefined,
                            amount: data_2.initiatePaymentContextWithExistingCustomer.amount,
                            currency: data_2.initiatePaymentContextWithExistingCustomer.currency_code,
                            metadata: {
                                resource_id: data_2.initiatePaymentContextWithExistingCustomer.resource_id,
                            },
                            capture_method: "manual",
                        }));
                        expect(result).toEqual(expect.objectContaining({
                            session_data: expect.any(Object),
                            update_requests: undefined,
                        }));
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on customer creation", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.initiatePayment(data_2.initiatePaymentContextWithWrongEmail)];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalledWith({
                            email: data_2.initiatePaymentContextWithWrongEmail.email,
                        });
                        expect(stripe_1.StripeMock.paymentIntents.create).not.toHaveBeenCalled();
                        expect(result).toEqual({
                            error: "An error occurred in initiatePayment when creating a Stripe customer",
                            code: "",
                            detail: "Error",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on payment intents creation", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.initiatePayment(data_2.initiatePaymentContextWithFailIntentCreation)];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalledWith({
                            email: data_2.initiatePaymentContextWithFailIntentCreation.email,
                        });
                        expect(stripe_1.StripeMock.paymentIntents.create).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.paymentIntents.create).toHaveBeenCalledWith(expect.objectContaining({
                            description: data_2.initiatePaymentContextWithFailIntentCreation.context
                                .payment_description,
                            amount: data_2.initiatePaymentContextWithFailIntentCreation.amount,
                            currency: data_2.initiatePaymentContextWithFailIntentCreation.currency_code,
                            metadata: {
                                resource_id: data_2.initiatePaymentContextWithFailIntentCreation.resource_id,
                            },
                            capture_method: "manual",
                        }));
                        expect(result).toEqual({
                            error: "An error occurred in InitiatePayment during the creation of the stripe payment intent",
                            code: "",
                            detail: "Error",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe("authorizePayment", function () {
        var _this = this;
        var stripeTest;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                scopedContainer = __assign({}, container);
                stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                return [2 /*return*/];
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should succeed", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.authorizePayment(data_2.authorizePaymentSuccessData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            data: data_2.authorizePaymentSuccessData,
                            status: medusa_1.PaymentSessionStatus.AUTHORIZED,
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe("cancelPayment", function () {
        var _this = this;
        var stripeTest;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                scopedContainer = __assign({}, container);
                stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                return [2 /*return*/];
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should succeed", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.cancelPayment(data_2.cancelPaymentSuccessData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on intent cancellation but still return the intent", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.cancelPayment(data_2.cancelPaymentPartiallyFailData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            id: stripe_1.PARTIALLY_FAIL_INTENT_ID,
                            status: types_1.ErrorIntentStatus.CANCELED,
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on intent cancellation", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.cancelPayment(data_2.cancelPaymentFailData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            error: "An error occurred in cancelPayment",
                            code: "",
                            detail: "Error",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe("capturePayment", function () {
        var _this = this;
        var stripeTest;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                scopedContainer = __assign({}, container);
                stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                return [2 /*return*/];
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should succeed", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.capturePayment(data_2.capturePaymentContextSuccessData.paymentSessionData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on intent capture but still return the intent", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.capturePayment(data_2.capturePaymentContextPartiallyFailData.paymentSessionData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            id: stripe_1.PARTIALLY_FAIL_INTENT_ID,
                            status: types_1.ErrorIntentStatus.SUCCEEDED,
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on intent capture", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.capturePayment(data_2.capturePaymentContextFailData.paymentSessionData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            error: "An error occurred in capturePayment",
                            code: "",
                            detail: "Error",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe("deletePayment", function () {
        var _this = this;
        var stripeTest;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                scopedContainer = __assign({}, container);
                stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                return [2 /*return*/];
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should succeed", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.cancelPayment(data_2.deletePaymentSuccessData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on intent cancellation but still return the intent", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.cancelPayment(data_2.deletePaymentPartiallyFailData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            id: stripe_1.PARTIALLY_FAIL_INTENT_ID,
                            status: types_1.ErrorIntentStatus.CANCELED,
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on intent cancellation", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.cancelPayment(data_2.deletePaymentFailData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            error: "An error occurred in cancelPayment",
                            code: "",
                            detail: "Error",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe("refundPayment", function () {
        var _this = this;
        var stripeTest;
        var refundAmount = 500;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                scopedContainer = __assign({}, container);
                stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                return [2 /*return*/];
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should succeed", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.refundPayment(data_2.refundPaymentSuccessData, refundAmount)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on refund creation", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.refundPayment(data_2.refundPaymentFailData, refundAmount)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            error: "An error occurred in refundPayment",
                            code: "",
                            detail: "Error",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe("retrievePayment", function () {
        var _this = this;
        var stripeTest;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                scopedContainer = __assign({}, container);
                stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                return [2 /*return*/];
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should succeed", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.retrievePayment(data_2.retrievePaymentSuccessData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            id: data_1.PaymentIntentDataByStatus.SUCCEEDED.id,
                            status: data_1.PaymentIntentDataByStatus.SUCCEEDED.status,
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail on refund creation", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.retrievePayment(data_2.retrievePaymentFailData)];
                    case 1:
                        result = _a.sent();
                        expect(result).toEqual({
                            error: "An error occurred in retrievePayment",
                            code: "",
                            detail: "Error",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe("updatePayment", function () {
        var _this = this;
        var stripeTest;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                scopedContainer = __assign({}, container);
                stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                return [2 /*return*/];
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should succeed to initiate a payment with an existing customer but no stripe id", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.updatePayment(data_2.updatePaymentContextWithExistingCustomer)];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalledWith({
                            email: data_2.updatePaymentContextWithExistingCustomer.email,
                        });
                        expect(stripe_1.StripeMock.paymentIntents.create).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.paymentIntents.create).toHaveBeenCalledWith(expect.objectContaining({
                            description: undefined,
                            amount: data_2.updatePaymentContextWithExistingCustomer.amount,
                            currency: data_2.updatePaymentContextWithExistingCustomer.currency_code,
                            metadata: {
                                resource_id: data_2.updatePaymentContextWithExistingCustomer.resource_id,
                            },
                            capture_method: "manual",
                        }));
                        expect(result).toEqual(expect.objectContaining({
                            session_data: expect.any(Object),
                            update_requests: {
                                customer_metadata: {
                                    stripe_id: stripe_1.STRIPE_ID,
                                },
                            },
                        }));
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail to initiate a payment with an existing customer but no stripe id", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.updatePayment(data_2.updatePaymentContextWithWrongEmail)];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.customers.create).toHaveBeenCalledWith({
                            email: data_2.updatePaymentContextWithWrongEmail.email,
                        });
                        expect(stripe_1.StripeMock.paymentIntents.create).not.toHaveBeenCalled();
                        expect(result).toEqual({
                            error: "An error occurred in updatePayment during the initiate of the new payment for the new customer",
                            code: "",
                            detail: "An error occurred in initiatePayment when creating a Stripe customer" +
                                os_1.EOL +
                                "Error",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should succeed but no update occurs when the amount did not changed", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.updatePayment(data_2.updatePaymentContextWithExistingCustomerStripeId)];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.paymentIntents.update).not.toHaveBeenCalled();
                        expect(result).not.toBeDefined();
                        return [2 /*return*/];
                }
            });
        }); });
        it("should succeed to update the intent with the new amount", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.updatePayment(data_2.updatePaymentContextWithDifferentAmount)];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.paymentIntents.update).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.paymentIntents.update).toHaveBeenCalledWith(data_2.updatePaymentContextWithDifferentAmount.paymentSessionData.id, {
                            amount: data_2.updatePaymentContextWithDifferentAmount.amount,
                        });
                        expect(result).toEqual({
                            session_data: expect.objectContaining({
                                amount: data_2.updatePaymentContextWithDifferentAmount.amount,
                            }),
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail to update the intent with the new amount", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.updatePayment(data_2.updatePaymentContextFailWithDifferentAmount)];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.paymentIntents.update).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.paymentIntents.update).toHaveBeenCalledWith(data_2.updatePaymentContextFailWithDifferentAmount.paymentSessionData.id, {
                            amount: data_2.updatePaymentContextFailWithDifferentAmount.amount,
                        });
                        expect(result).toEqual({
                            error: "An error occurred in updatePayment",
                            code: "",
                            detail: "Error",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
    describe("updatePaymentData", function () {
        var _this = this;
        var stripeTest;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            var scopedContainer;
            return __generator(this, function (_a) {
                scopedContainer = __assign({}, container);
                stripeTest = new stripe_test_1.StripeTest(scopedContainer, { api_key: "test" });
                return [2 /*return*/];
            });
        }); });
        beforeEach(function () {
            jest.clearAllMocks();
        });
        it("should succeed to update the payment data", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.updatePaymentData(data_2.updatePaymentDataWithoutAmountData.sessionId, __assign(__assign({}, data_2.updatePaymentDataWithoutAmountData), { sessionId: undefined }))];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.paymentIntents.update).toHaveBeenCalled();
                        expect(stripe_1.StripeMock.paymentIntents.update).toHaveBeenCalledWith(data_2.updatePaymentDataWithoutAmountData.sessionId, {
                            customProp: data_2.updatePaymentDataWithoutAmountData.customProp,
                        });
                        expect(result).toEqual(expect.objectContaining({
                            customProp: data_2.updatePaymentDataWithoutAmountData.customProp,
                        }));
                        return [2 /*return*/];
                }
            });
        }); });
        it("should fail to update the payment data if the amount is present", function () { return __awaiter(_this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, stripeTest.updatePaymentData(data_2.updatePaymentDataWithAmountData.sessionId, __assign(__assign({}, data_2.updatePaymentDataWithAmountData), { sessionId: undefined }))];
                    case 1:
                        result = _a.sent();
                        expect(stripe_1.StripeMock.paymentIntents.update).not.toHaveBeenCalled();
                        expect(result).toEqual({
                            error: "An error occurred in updatePaymentData",
                            code: undefined,
                            detail: "Cannot update amount, use updatePayment instead",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    });
});
